Command framework revamp

jnwhiteh opened the pull request:
Implements /give, /tp

jnwhiteh commented on src/chunkymonkey/command/command.go:
```diff
@@ -1,6 +1,9 @@
+// A CommandCallback takes the name of the player invoking the command, any
+// text supplied after the trigger for the command, and an interface via which
+// game-wide 'actions' can be taken.
+type CommandCallback func(string, string, ICommandHandler)
```
Are we specifying parameter names in function types? I'm happy to add them, since its not entirely obvious at all what this function is supposed to do, although the comment does describe the parameters.

huin replied:
I think that parameter names are possibly the most important documentation about the parameters. The type is obviously required and useful, but the names imply some extra meaning. In this case it could gain a lot from having the names.

jnwhiteh commented on src/chunkymonkey/command/commands.go:
```diff
@@ -111,47 +124,67 @@ func cmdHelp(message string, cmdFramework *CommandFramework, cmdHandler ICommand
 	}
 
-	cmdHandler.GiveItem(itemId, quantity, data)
+	// TODO: How can we get the name of this item without a dependency loop?
```
I'd love any input on how we can do this, it escapes me right now.

huin replied:
I think my upcoming counter-pull request for you to look at might help there.

jnwhiteh replied:
Definitely, that's my goal.

On Tue, Jul 26, 2011 at 9:00 AM, huin
reply@reply.github.com
wrote:

> >                       return
> >               }
> >       }
> > -     cmdHandler.GiveItem(itemId, quantity, data)
> > -     // TODO: How can we get the name of this item without a dependency loop?
> 
> I think my upcoming counter-pull request for you to look at might help there.
> 
> ## 
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/huin/chunkymonkey/pull/77/files#r74029

jnwhiteh committed:
Merge pull request #1 from huin/jnwhiteh_commands

Commands dependency change

jnwhiteh commented:
This now includes your work. We can either merge it now, or we can wait until I've done the IGame revamp. Either way, I'll see to that tonight, if I can.

huin commented:
I'm happy either way, although I guess I have a slight bias to having
a complete revamp for merging.

jnwhiteh committed:
Revamp command framework to use IGame, IPlayerClient

Currently the /tp behaviour is a bit odd (there is full collision, so the person being teleported to can end up moving). Tests are currently not passing due to lack of gomoc-fu, but I plan to update them in the next commit.

jnwhiteh committed:
Tests compile, but are not all passing yet

jnwhiteh committed:
Add missing player_client file
